states/state.go

```diff
@@ -223,6 +224,22 @@ func (s *State) ProviderAddrs() []addrs.AbsProviderConfig {
 	return ret
 }
 
+// ProviderRequirements returns a description of all of the providers that
+// are required to work with the receiving state.
+//
+// Because the state does not track specific version information for providers,
+// the requirements returned by this method will always be unconstrained.
+// The result should usually be merged with a Requirements derived from the
+// current configuration in order to apply some constraints.
+func (s *State) ProviderRequirements() getproviders.Requirements {
+	configAddrs := s.ProviderAddrs()
```

Comment:
Are we getting any implied providers used by resources from state somewhere? if I'm reading this correctly, it only gathers configured providers.

Reply:
Here's what I *think* is the analogous current code, which iterates over every resource in state:

https://github.com/hashicorp/terraform/blob/369b0560b9bc86cadaefa0f23dd462e8929a94da/terraform/module_dependencies.go#L198-L211

Reply:
It looks to me like the `ProviderAddrs` method above is doing something functionality equivalent to this:

```go
	for _, ms := range s.Modules {
		for _, rc := range ms.Resources {
			m[rc.ProviderConfig.String()] = rc.ProviderConfig
		}
	}
```

Does that seem equivalent to you too, or am I missing something? :thinking:

Reply:
Nope, that looks exactly right. I'd _thought_ I'd seen it only looping over the module's provider configs, and clearly I was mistaken! Thank you!